rmw_zenoh_cpp/src/detail/logging_macros.hpp

```diff
@@ -0,0 +1,37 @@
+// Copyright 2024 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DETAIL__LOGGING_MACROS_HPP_
+#define DETAIL__LOGGING_MACROS_HPP_
+
+#include "logging.hpp"
+
+///=============================================================================
+// We define custom logging marcos to log to console as it was discovered
+// that relying on RCUTILS_LOG_X_NAMED functions also end up logging over /rosout
+// which can lead to deadlocks in rmw_zenoh especially when multiple threads
+// invoke GraphCache::parse_put() and GraphCache::parse_del() functions.
+// See https://github.com/ros2/rmw_zenoh/issues/182 for more details.
+// Note: Consider adding a set_rmw_log_level() API to rmw which can override the
+// default log level at the RMW layer.
```

Comment:
This actually already exists: https://github.com/ros2/rmw_zenoh/blob/0eb456032355a2be9965e5171e09286046c8a0a1/rmw_zenoh_cpp/src/rmw_zenoh.cpp#L3874-L3879 .

We just don't currently implement it, and I don't believe `rcl` or anything else currently calls it.

Reply:
Oh nice! I've implemented it in https://github.com/ros2/rmw_zenoh/pull/223/commits/73af6701512578f96103e2060aa5bd6ce3ebb2d7 since connext and fastrtps implement it too.